File: old/julia_version/simulate.jl
#!/usr/bin/env julia

"""
Main simulation driver for RBF-FD Navier-Stokes solver.

This is the Julia equivalent of simulate.m - the main executable script
that orchestrates the complete simulation workflow.
"""

using Pkg
Pkg.activate(@__DIR__)

include("RBFNS.jl")
using .RBFNS
using LinearAlgebra, SparseArrays
using Printf

function main()
    println("🌊 Julia RBF-FD Navier-Stokes Solver")
    println("=" ^ 50)
    
    # 1) Load Configuration
    println("\n📋 1. Loading configuration...")
    cfg = default_config()
    println("   Reynolds number: $(cfg.reynolds_number)")
    println("   Time step: $(cfg.time_step)")
    println("   Geometry: $(cfg.geometry_type)")
    println("   Domain: [$(cfg.x_min), $(cfg.x_max)] × [$(cfg.y_min), $(cfg.y_max)]")
    
    # Determine number of time steps
    Nt = cfg.num_time_steps
    if haskey(ENV, "CI") && ENV["CI"] == "true"
        Nt = cfg.num_time_steps_ci
        println("   CI mode detected: using $Nt time steps")
    end
    
    # 2) Generate geometry
    println("\n🏗️  2. Generating geometry...")
    G = build_geometry(cfg)
    
    println("   Interior velocity nodes: $(size(G.xy, 1))")
    println("   Interior pressure nodes: $(size(G.xy_s, 1))")
    println("   Boundary nodes:")
    println("     Inlet: $(size(G.boundary_in, 1)) (V), $(size(G.boundary_in_s, 1)) (P)")
    println("     Outlet: $(size(G.boundary_out, 1)) (V), $(size(G.boundary_out_s, 1)) (P)")
    println("     Walls: $(size(G.boundary_y, 1)) (V), $(size(G.boundary_y_s, 1)) (P)")
    println("     Obstacle: $(size(G.boundary_obs, 1)) (V), $(size(G.boundary_obs_s, 1)) (P)")
    
    # 3) Build stencils for RBF-FD method
    println("\n🎯 3. Building RBF-FD stencils...")
    
    # Combine grids for stencil construction
    xy1 = vcat(G.xy, G.boundary_in, G.boundary_out, G.boundary_y, G.boundary_obs)
    xy1_s = vcat(G.xy_s, G.boundary_in_s, G.boundary_out_s, G.boundary_y_s, G.boundary_obs_s)
    
    println("   Total V-grid nodes: $(size(xy1, 1))")
    println("   Total P-grid nodes: $(size(xy1_s, 1))")
    
    # Build nearest neighbor indices
    k_main = cfg.stencil_size_main
    Nearest_Idx = nearest_interp(xy1, xy1, k_main)
    Nearest_Idx_s = nearest_interp(xy1_s, xy1_s, k_main)
    
    println("   Stencil size: $k_main")
    println("   Built neighbor indices")
    
    # 4) Build RBF-FD operators
    println("\n🔧 4. Building RBF-FD operators...")
    
    # Velocity grid operators
    Dx, Dy, L0, Dxx, Dyy, Dxy = rbf_phs_fd_all(xy1, xy1, Nearest_Idx, 
                                                k_main, cfg.order_main, cfg.poly_degree_main)
    
    # Pressure grid operators  
    Dx_s, Dy_s, L_s, Dxx_s, Dyy_s, Dxy_s = rbf_phs_fd_all(xy1_s, xy1_s, Nearest_Idx_s,
                                                           k_main, cfg.order_main, cfg.poly_degree_main)
    
    # Inter-grid operators (P-grid to V-grid and vice versa)
    Nearest_Idx_12 = nearest_interp(xy1, xy1_s, k_main)
    Nearest_Idx_21 = nearest_interp(xy1_s, xy1, k_main)
    
    D0_12_x, D0_12_y, _, _, _, _ = rbf_phs_fd_all(xy1, xy1_s, Nearest_Idx_12,
                                                  k_main, cfg.order_main, cfg.poly_degree_main)
    D0_21_x, D0_21_y, _, _, _, _ = rbf_phs_fd_all(xy1_s, xy1, Nearest_Idx_21,
                                                  k_main, cfg.order_main, cfg.poly_degree_main)
    
    println("   Built velocity operators: $(size(Dx)) → $(size(Dy))")
    println("   Built pressure operators: $(size(L_s))")
    println("   Built inter-grid operators")
    
    # 5) Setup boundary condition operators (simplified)
    println("\n🚧 5. Setting up boundary conditions...")
    
    n_interior = size(G.xy, 1)
    n_boundary_y = size(G.boundary_y, 1)
    n_boundary_out = size(G.boundary_out, 1)
    n_total = size(xy1, 1)
    
    # Create simplified boundary operators (placeholders)
    Dy_b_0 = spzeros(n_boundary_y, n_total)
    Dx_b_0 = spzeros(n_boundary_out, n_total)
    
    # For wall boundaries: du/dy = 0 (simplified as identity for now)
    if n_boundary_y > 0
        wall_indices = n_interior + 1:n_interior + n_boundary_y
        for (i, idx) in enumerate(wall_indices)
            Dy_b_0[i, idx] = 1.0
        end
    end
    
    # For outlet boundaries: du/dx = 0 (simplified as identity for now)  
    if n_boundary_out > 0
        outlet_indices = n_interior + n_boundary_y + 1:n_interior + n_boundary_y + n_boundary_out
        for (i, idx) in enumerate(outlet_indices)
            Dx_b_0[i, idx] = 1.0
        end
    end
    
    println("   Wall BC operator: $(size(Dy_b_0))")
    println("   Outlet BC operator: $(size(Dx_b_0))")
    
    # 6) Build velocity solvers
    println("\n⚡ 6. Precomputing velocity solvers...")
    
    L_u_inv, L_v_inv = precompute_velocity_solvers(L0, Dy_b_0, Dx_b_0,
                                                   n_interior, n_boundary_y, n_boundary_out, cfg)
    
    println("   Velocity solvers precomputed")
    
    # 7) Build pressure system (MATLAB-style with proper boundary conditions)
    println("\n💨 7. Building pressure system...")
    
    n_interior_s = size(G.xy_s, 1)
    boundary_s_nodes = vcat(G.boundary_y_s, G.boundary_out_s, G.boundary_in_s, G.boundary_obs_s)
    n_boundary_s = size(boundary_s_nodes, 1)
    
    # Start with Laplacian for interior nodes (matching MATLAB line 87)
    L_pressure = copy(L_s[1:n_interior_s, :])
    
    # Build STABLE boundary condition operators (simplified but numerically robust)
    # Use homogeneous Neumann BCs (dp/dn = 0) implemented as simplified zero-derivative constraints
    L_bc_total = spzeros(n_boundary_s, size(xy1_s, 1))
    
    # Simple approach: enforce p_boundary ≈ p_interior_average (weak Neumann approximation)
    boundary_idx = 1
    for i in 1:size(G.boundary_y_s, 1)
        # Wall: dp/dy ≈ 0 → p_wall ≈ p_interior 
        L_bc_total[boundary_idx, n_interior_s + boundary_idx] = 1.0
        boundary_idx += 1
    end
    
    for i in 1:size(G.boundary_out_s, 1)
        # Outlet: dp/dx ≈ 0 → p_outlet ≈ p_interior
        L_bc_total[boundary_idx, n_interior_s + boundary_idx] = 1.0  
        boundary_idx += 1
    end
    
    for i in 1:size(G.boundary_in_s, 1)
        # Inlet: dp/dx ≈ 0 → p_inlet ≈ p_interior
        L_bc_total[boundary_idx, n_interior_s + boundary_idx] = 1.0
        boundary_idx += 1
    end
    
    for i in 1:size(G.boundary_obs_s, 1)
        # Obstacle: dp/dn ≈ 0 → p_obstacle ≈ p_interior
        L_bc_total[boundary_idx, n_interior_s + boundary_idx] = 1.0
        boundary_idx += 1
    end
    
    # Assemble: [Laplacian at interior; BCs at boundary] (MATLAB line 87)
    L_pressure = vcat(L_pressure, L_bc_total)
    
    # Add regularization exactly like MATLAB (line 91)
    # [L1 [ones(length(xy_s),1); zeros(length(boundary_s),1)]; [ones(1,length(xy_s)) zeros(1,length(boundary_s))] 0]
    reg_col = vcat(ones(n_interior_s), zeros(n_boundary_s))
    reg_row = hcat(ones(1, n_interior_s), zeros(1, n_boundary_s))
    
    L_pressure = hcat(L_pressure, reg_col)
    L_pressure = vcat(L_pressure, hcat(reg_row, [0.0]))
    
    L_inv_s = build_pressure_system(sparse(L_pressure))
    
    println("   Pressure system: $(size(L_pressure))")
    
    # 8) Create simplified inter-grid and boundary operators
    println("\n🔗 8. Setting up remaining operators...")
    
    # Simplified operators for demonstration
    Dy_b = spzeros(n_boundary_y, n_total)
    Dy_b_1 = ones(n_boundary_y)
    D0_21_x_obs = spzeros(size(G.boundary_obs, 1), size(xy1_s, 1))
    D0_21_y_obs = spzeros(size(G.boundary_obs, 1), size(xy1_s, 1))
    
    # Boundary size parameters
    L_B = size(G.boundary_in, 1) + size(G.boundary_obs, 1)
    L_B_obs = size(G.boundary_obs, 1)
    L_W = n_interior
    L_B_y = n_boundary_y
    
    # Compute L_B_S as total pressure boundary nodes (matching MATLAB's length(boundary_s))
    L_B_S = n_boundary_s
    
    println("   Boundary parameters: L_B=$L_B, L_W=$L_W, L_B_y=$L_B_y")
    
    # 9) Initialize simulation state
    println("\n🎬 9. Initializing simulation...")
    
    n_velocity_nodes = size(xy1, 1)
    W = zeros(2 * n_velocity_nodes, Nt + 1)  # [u; v] for all time steps
    p0 = zeros(size(xy1_s, 1))  # Initial pressure
    
    # Set initial conditions (zero everywhere for simplicity)
    W[:, 1] .= 0.0
    
    # Set inlet boundary conditions (unit velocity in x-direction)
    if size(G.boundary_in, 1) > 0
        inlet_start = n_interior + 1
        inlet_end = n_interior + size(G.boundary_in, 1)
        W[inlet_start:inlet_end, 1] .= 1.0  # u = 1 at inlet
    end
    
    println("   Initial state set")
    println("   Velocity vector size: $(size(W, 1))")
    
    # 10) Main time-stepping loop
    println("\n🔄 10. Starting time integration...")
    println("    Total time steps: $Nt")
    
    for j in 1:Nt
        if j % max(1, Nt ÷ 10) == 0
            @printf("    Step %d/%d (%.1f%%)\n", j, Nt, 100*j/Nt)
        end
        
        if j < 3
            # First few steps: use simple first-order scheme for stability
            W[:, j+1] = W[:, j]
        else
            # Main fractional step algorithm
            try
                W[:, j+1], p0 = ns2d_fractional_step_phs(
                    cfg.time_step, cfg.viscosity,
                    W[:, j-1], W[:, j],
                    Dy, Dx, L_inv_s, L_u_inv, L_v_inv, L0,
                    L_B, L_B_obs, L_W, L_B_y, L_B_S,
                    D0_12_x, D0_12_y, D0_21_x, D0_21_y,
                    Dy_b, Dy_b_1, D0_21_x_obs, D0_21_y_obs, p0, W[:, 1], cfg
                )
            catch e
                println("    ⚠️  Error in time step $j: $e")
                println("    Stopping simulation")
                break
            end
        end
        
        # Check for numerical instability
        if any(isnan.(W[:, j+1]))
            println("    ⚠️  NaN detected at step $j - simulation unstable")
            break
        end
    end
    
    W0 = W[:, end]
    
    # 11) Final results
    println("\n📊 11. Simulation completed!")
    println("    Final velocity norm: $(norm(W0))")
    println("    Maximum velocity: $(maximum(abs.(W0)))")
    
    u_final = W0[1:n_velocity_nodes]
    v_final = W0[n_velocity_nodes+1:end]
    
    println("    u-velocity range: [$(minimum(u_final)), $(maximum(u_final))]")
    println("    v-velocity range: [$(minimum(v_final)), $(maximum(v_final))]")
    
    println("\n✅ Julia RBF-FD simulation complete!")
    
    return W, xy1, G, cfg
end

# Run simulation if this file is executed directly
if abspath(PROGRAM_FILE) == @__FILE__
    try
        W, xy1, G, cfg = main()
        println("\n🎉 Success! Results stored in variables W, xy1, G, cfg")
    catch e
        println("\n❌ Simulation failed with error:")
        println(e)
        if isa(e, InterruptException)
            println("Simulation interrupted by user")
        else
            # Print stack trace for debugging
            for (exc, bt) in Base.catch_stack()
                showerror(stdout, exc, bt)
                println()
            end
        end
    end
end

File: simulate.m
clc;
clear;

% Add required paths for src functions and lib dependencies
scriptDir = fileparts(mfilename('fullpath'));

% Add src directory for supporting functions (config.m is now in main directory)
srcPath = fullfile(scriptDir, 'src');
if exist(srcPath, 'dir')
  addpath(srcPath);
  % Also add geometry subdirectory
  geometryPath = fullfile(srcPath, 'geometry');
  if exist(geometryPath, 'dir')
    addpath(geometryPath);
  end
end

% Load configuration parameters
% Can be overridden by setting CONFIG_FUNC and GEOMETRY_TYPE environment variables
config_func_name = getenv('CONFIG_FUNC');
geometry_type = getenv('GEOMETRY_TYPE');

if isempty(config_func_name)
  config_func_name = 'config';
end

if isempty(geometry_type)
  % Default behavior - call config function without arguments
  cfg = config();
else
  % Pass geometry type to config function
  cfg = config(geometry_type);
end

% Make config globally accessible for diagnostic functions
global cfg

%% 1) Setup environment and dependencies
[doPlot, isCI, isTest, Nt] = setup_environment(cfg, scriptDir);

% Extract domain parameters - define computational domain boundaries
x_min = cfg.domain.x_min;  % Left boundary of domain
x_max = cfg.domain.x_max;  % Right boundary of domain
y_min = cfg.domain.y_min;  % Bottom boundary of domain
y_max = cfg.domain.y_max;  % Top boundary of domain

%% 2) Generate geometry
G = build_geometry(cfg);

% Extract mesh data from geometry structure
xy = G.xy;                    % Interior velocity nodes
xy_s = G.xy_s;               % Interior pressure nodes
xt = G.xt;                   % Triangle connectivity

% Extract velocity grid boundaries
boundary_in = G.boundary_in;     % Inlet boundary nodes (V-grid)
boundary_out = G.boundary_out;   % Outlet boundary nodes (V-grid)
boundary_y = G.boundary_y;       % Wall boundary nodes (V-grid)
boundary_obs = G.boundary_obs;   % Obstacle boundary nodes (V-grid)

% Extract pressure grid boundaries
boundary_in_s = G.boundary_in_s;   % Inlet boundary nodes (P-grid)
boundary_out_s = G.boundary_out_s; % Outlet boundary nodes (P-grid)
boundary_y_s = G.boundary_y_s;     % Wall boundary nodes (P-grid)
boundary_obs_s = G.boundary_obs_s; % Obstacle boundary nodes (P-grid)

% Extract geometry-specific parameters (used for distance calculations)
if isfield(G, 'radius')
  radius = G.radius;           % Cylinder radius
  obs_char_length = radius;    % Characteristic length for distance calculations
elseif isfield(G, 'a') && isfield(G, 'b')
  a = G.a;                     % Ellipse semi-major axis
  b = G.b;                     % Ellipse semi-minor axis
  obs_char_length = min(a, b); % Use minimum axis as characteristic length
elseif isfield(G, 'rect_width') && isfield(G, 'rect_height')
  rect_width = G.rect_width;   % Rectangle width
  rect_height = G.rect_height; % Rectangle height
  obs_char_length = min(rect_width, rect_height) / 2; % Use half of smaller dimension
elseif isfield(G, 'chord_length') && isfield(G, 'naca_digits')
  chord_length = G.chord_length; % Airfoil chord length
  naca_digits = G.naca_digits;   % NACA 4-digit parameters
  % Calculate maximum thickness as characteristic length
  thickness_percent = naca_digits(3) * 10 + naca_digits(4); % Thickness as percentage
  max_thickness = (thickness_percent / 100) * chord_length;  % Maximum thickness in units
  obs_char_length = max_thickness / 2; % Use half of maximum thickness
elseif isfield(G, 'obstacles') && isfield(G, 'num_obstacles')
  % Multi-obstacle geometry: use minimum characteristic length of all obstacles
  obs_char_length = inf;
  for i = 1:G.num_obstacles
    obs = G.obstacles(i);
    switch lower(obs.type)
      case 'cylinder'
        char_len = obs.params.radius;
      case 'ellipse'
        char_len = min(obs.params.a, obs.params.b);
      case 'rectangle'
        char_len = min(obs.params.width, obs.params.height) / 2;
      case 'airfoil'
        thickness_percent = obs.params.naca_digits(3) * 10 + obs.params.naca_digits(4);
        max_thickness = (thickness_percent / 100) * obs.params.chord_length;
        char_len = max_thickness / 2;
      otherwise
        char_len = 0.5; % Default fallback
    end
    obs_char_length = min(obs_char_length, char_len);
  end
else
  error('Geometry structure missing expected parameters');
end

% Combine boundary nodes in specific order for compatibility
boundary_s = [boundary_y_s; boundary_out_s; boundary_in_s; boundary_obs_s];
boundary = [boundary_y; boundary_out; boundary_in; boundary_obs];

% Initial mesh visualization (only in non-CI mode)
if doPlot && ~isCI && ~isTest
  fprintf('Creating mesh visualization...\n');
  fig = figure('Name', 'Multi-Obstacle Mesh', 'Position', [100, 100, 1000, 600]);

  % Plot interior nodes (sample for performance if needed)
  if size(xy, 1) > 10000
    % Sample for very large meshes
    sample_rate = max(1, floor(size(xy, 1) / 8000));
    scatter(xy(1:sample_rate:end, 1), xy(1:sample_rate:end, 2), 4, 'k.', 'DisplayName', ...
            sprintf('Interior nodes (sampled %d/%d)', length(1:sample_rate:size(xy, 1)), size(xy, 1)));
  else
    % Show all interior nodes for reasonable mesh sizes
    scatter(xy(:, 1), xy(:, 2), 3, 'k.', 'DisplayName', sprintf('Interior nodes (%d)', size(xy, 1)));
  end
  hold on;

  % Plot pressure grid interior nodes (lighter color)
  if size(xy_s, 1) > 5000
    % Sample pressure nodes for very large meshes
    p_sample_rate = max(1, floor(size(xy_s, 1) / 4000));
    scatter(xy_s(1:p_sample_rate:end, 1), xy_s(1:p_sample_rate:end, 2), 2, [0.7 0.7 0.7], '.', ...
            'DisplayName', sprintf('Pressure nodes (sampled %d/%d)', length(1:p_sample_rate:size(xy_s, 1)), size(xy_s, 1)));
  else
    scatter(xy_s(:, 1), xy_s(:, 2), 2, [0.7 0.7 0.7], '.', ...
            'DisplayName', sprintf('Pressure nodes (%d)', size(xy_s, 1)));
  end

  % Plot boundary nodes with different colors and markers
  scatter(boundary_in(:, 1), boundary_in(:, 2), 25, 'b', 's', 'filled', 'DisplayName', sprintf('Inlet (%d)', length(boundary_in)));
  scatter(boundary_y(:, 1), boundary_y(:, 2), 25, 'r', '^', 'filled', 'DisplayName', sprintf('Walls (%d)', length(boundary_y)));
  scatter(boundary_out(:, 1), boundary_out(:, 2), 25, 'g', 'v', 'filled', 'DisplayName', sprintf('Outlet (%d)', length(boundary_out)));

  % Color-code obstacle boundaries if multi-obstacle
  if isfield(G, 'boundary_obs_ids') && isfield(G, 'num_obstacles')
    colors = {[1 0 1], [0 1 1], [1 1 0], [0.8 0.4 0], [0.6 0.2 0.8], [0.2 0.8 0.2], [0.8 0.2 0.2]};
    for obs_id = 1:G.num_obstacles
      obs_mask = G.boundary_obs_ids == obs_id;
      if any(obs_mask)
        color_idx = mod(obs_id - 1, length(colors)) + 1;
        color = colors{color_idx};
        scatter(G.boundary_obs_s(obs_mask, 1), G.boundary_obs_s(obs_mask, 2), ...
                35, color, 'o', 'filled', 'LineWidth', 1, 'MarkerEdgeColor', 'k', ...
                'DisplayName', sprintf('Obstacle %d (%d pts)', obs_id, sum(obs_mask)));
      end
    end
  else
    scatter(boundary_obs(:, 1), boundary_obs(:, 2), 35, [1 0 1], 'o', 'filled', ...
            'LineWidth', 1, 'MarkerEdgeColor', 'k', ...
            'DisplayName', sprintf('Obstacle (%d pts)', length(boundary_obs)));
  end

  axis equal;
  grid on;

  % Calculate total node counts
  total_v_nodes = size(xy, 1) + length(boundary_in) + length(boundary_y) + length(boundary_out) + length(boundary_obs);
  total_p_nodes = size(xy_s, 1) + length(G.boundary_y_s) + length(G.boundary_out_s) + length(G.boundary_in_s) + length(G.boundary_obs_s);
  if isfield(G, 'num_obstacles')
    num_obstacles = G.num_obstacles;
  else
    num_obstacles = 1; % Single obstacle case
  end

  title(sprintf('Multi-Obstacle Mesh: %d obstacles, V-grid: %d nodes, P-grid: %d nodes', ...
                num_obstacles, total_v_nodes, total_p_nodes));
  xlabel('x');
  ylabel('y');
  legend('Location', 'best', 'FontSize', 8);

  % Save figure and display
  try
    saveas(fig, 'multi_obstacle_mesh.png');
    fprintf('Mesh visualization saved as multi_obstacle_mesh.png\n');
  catch
    % Ignore save errors
  end

  % Force figure to front and pause briefly to ensure visibility
  figure(fig);
  drawnow;
  fprintf('Mesh visualization displayed. Continuing with simulation...\n');
  pause(1); % Brief pause to let user see the mesh
elseif isCI || isTest
  fprintf('Skipping mesh visualization (CI/test mode)\n');
end

% Combine interior and boundary nodes to create complete grids
xy1 = [xy; boundary];      % Complete velocity grid (V-grid): interior + boundary
xy1_s = [xy_s; boundary_s]; % Complete pressure grid (P-grid): interior + boundary

% Extract coordinates for convenience
x1 = xy1(:, 1);   % x-coordinates of velocity nodes
y1 = xy1(:, 2);   % y-coordinates of velocity nodes
x1_s = xy1_s(:, 1); % x-coordinates of pressure nodes
y1_s = xy1_s(:, 2); % y-coordinates of pressure nodes
x0 = xy(:, 1);    % x-coordinates of interior velocity nodes only
y0 = xy(:, 2);    % y-coordinates of interior velocity nodes only
x0_s = xy_s(:, 1); % x-coordinates of interior pressure nodes only
y0_s = xy_s(:, 2); % y-coordinates of interior pressure nodes only

% Define distance thresholds for special RBF treatment near boundaries
x_min_dist = cfg.distances.x_min;  % Distance threshold from left boundary
x_max_dist = cfg.distances.x_max;  % Distance threshold from right boundary
y_min_dist = cfg.distances.y_min;  % Distance threshold from bottom boundary
y_max_dist = cfg.distances.y_max;  % Distance threshold from top boundary
r_dist = cfg.mesh.refine_a2 * cfg.mesh.edge_multiplier;  % Distance threshold from cylinder

%% 3) Build stencils for RBF-FD method
S = build_stencils(xy1, xy1_s, xy, xy_s, G, cfg);

%% 4) Build pressure system and boundary conditions
P = build_pressure_system(G, xy_s, xy1_s, cfg);
L_inv_s = P.L_inv_s;
D0_21_x_obs = P.D0_21_x_obs;
D0_21_y_obs = P.D0_21_y_obs;

%% 5) Build inter-grid operators (P-grid ↔ V-grid)
[D0_21_x, D0_21_y, D0_12_x, D0_12_y] = build_intergrid_ops(G, xy, xy1, xy_s, xy1_s, S, cfg);

%% 6) Build velocity operators and boundary conditions
Vops = build_velocity_operators(G, xy, xy1, boundary_y, boundary_out, S, cfg, cfg.distances);

Dx = Vops.Dx;
Dy = Vops.Dy;
L0 = Vops.L0;
Dy_b_0 = Vops.Dy_b_0;
Dx_b_0 = Vops.Dx_b_0;
Dy_b = Vops.Dy_b;
Dy_b_1 = Vops.Dy_b_1;

%% 7) Precompute velocity system solvers
[L_u_inv, L_v_inv] = precompute_velocity_solvers(L0, Dy_b_0, Dx_b_0, xy, boundary_y, boundary_out, cfg);

%% 8) Setup time-stepping parameters
nu = cfg.simulation.viscosity;  % Kinematic viscosity (1/Reynolds number)
dt = cfg.simulation.time_step;  % Time step size

%% 8.1) Precompute mesh spacing stats for CFL-like diagnostics
h_min = NaN;
h_med = NaN;
if cfg.logging.enable
  try
    [~, D2] = knnsearch(xy, xy, 2);  % 2nd neighbor distance on interior V-grid
    if size(D2, 2) >= 2
      d2 = D2(:, 2);
      h_min = min(d2);
      h_med = median(d2);
      fprintf('[diag] spacing: h_min=%.3g, h_med=%.3g (V-grid interior)\n', h_min, h_med);
    end
  catch ME
    warning('spacing precomp failed: %s', ME.message);
  end
end

%% 9) Initialize simulation state
[W, p0] = init_state(xy1, xy1_s, boundary_obs, Nt);

% Define useful index lengths for boundary handling
L_B = length(boundary_obs) + length(boundary_in);   % Total special boundaries
L_B_y = length(boundary_y);   % Wall boundaries
L_W = length(xy);             % Interior velocity nodes
L_B_obs = length(boundary_obs);   % Obstacle boundary nodes

% Define pressure boundary indices for easy access
Idx_y_s = length(xy_s) + 1:length(xy_s) + length(boundary_y_s);    % Wall pressure indices
Idx_in_s = length(xy1_s) + 1 - length(boundary_obs_s) - length(boundary_in_s): ...
           length(xy1_s) - length(boundary_obs_s);  % Inlet pressure indices
Idx_out_s = length(xy_s) + 1 + length(boundary_y_s): ...
            length(xy_s) + length(boundary_out_s) + length(boundary_y_s);  % Outlet pressure indices

%% 10) Main time-stepping loop: Fractional Step Method
for j = 1:Nt
  if cfg.simulation.show_progress && ~isCI && ~isTest
    disp(['Time step j = ' num2str(j)]);
  end

  try
    % Use different schemes for startup vs main integration
    if j < 3
      % First few steps: use simple first-order scheme for stability
      W(:, j + 1) = W(:, j);  % Copy previous solution for startup
    else
      % Main fractional step algorithm implemented in NS_2d_fractional_step_PHS
      % This performs: 1) Advection-diffusion step with Adams-Bashforth + Crank-Nicolson
      %                2) Pressure correction step to enforce incompressibility
      %                3) Velocity correction using pressure gradient
      [W(:, j + 1), p0] = NS_2d_fractional_step_PHS(dt, nu, W(:, j - 1), W(:, j), ...
                                                    Dy, Dx, L_inv_s, L_u_inv, L_v_inv, ...
                                                    L0, L_B, L_B_obs, L_W, L_B_y, ...
                                                    length(boundary_s), D0_12_x, D0_12_y, ...
                                                    D0_21_x, D0_21_y, Dy_b, Dy_b_1, ...
                                                    D0_21_x_obs, D0_21_y_obs, p0, W(:, 1));
    end
  catch ME
    warning('Step %d threw error: %s', j, ME.message);
    if cfg.logging.snapshot_on_nan
      save_debug_snapshot(cfg, j, xy1, W, p0, dt, Dx, Dy, D0_12_x, D0_12_y, G);
    end
    rethrow(ME);
  end

  % Diagnostics every N steps
  if cfg.logging.enable && (mod(j, cfg.logging.step_frequency) == 0 || j <= 3)
    log_step_stats(j, W(:, j), W(:, j + 1), dt, Dx, Dy, D0_12_x, D0_12_y, xy1, h_min, h_med, cfg);
  end

  % Instability detection and snapshot
  if any(~isfinite(W(:, j + 1)))
    bad_idx = find(~isfinite(W(:, j + 1)), 1, 'first');
    [loc_str, pt] = classify_node_idx(bad_idx, L_W, boundary_y, boundary_out, boundary_in, boundary_obs, xy1);

    fprintf('\n*** SIMULATION INSTABILITY DETECTED ***\n');
    fprintf('Step: %d, Index: %d, Location: %s at (%.3g, %.3g)\n', j, bad_idx, loc_str, pt(1), pt(2));

    % Perform detailed instability analysis
    analyze_instability_cause(j, W, p0, bad_idx, loc_str, pt, dt, Dx, Dy, D0_12_x, D0_12_y, xy1, h_min, h_med, cfg);

    if cfg.logging.snapshot_on_nan
      save_debug_snapshot(cfg, j, xy1, W, p0, dt, Dx, Dy, D0_12_x, D0_12_y, G);
    end
    break
  end
end

% Extract final velocity field for potential continuation
W0 = W(:, end);

%% 11) Visualization of final results
visualize_final(cfg, doPlot, xy1, W, Nt, x_min, x_max, y_min, y_max, Dx, Dy);
